### PR TITLE
V0.5

### PR DESCRIPTION
**Pull request by srikanth-scale**
Working on new python API please suggest docs/improvements/cleanups 

ty

### CODE REVIEW

**Comment by srikanth-scale** [comment]
Working well, tested locally but extremely ugly -> @Nastia39 need your suggestions to cleanup everything

File: nucleus/__init__.py [hunk 1]
```diff
@@ -175,20 +207,46 @@ def populate_dataset(
...
+
+        for batch in tqdm_remote_batches:
+            payload = construct_append_payload(batch)
+            responses = self._process_append_requests(dataset_id, payload, batch_size, batch_size)
+            async_responses.extend(responses)
+
         for response in async_responses:
             agg_response.update_response(response.json())
 
+        print(agg_response.json())
```
Comment (rkaplan) [inline nucleus/__init__.py:249]: Can we remove these print statements before merging

File: nucleus/__init__.py [hunk 2]
```diff
@@ -294,17 +354,41 @@ def exception_handler(request, exception):
...
+            payload = construct_box_annotation_payload(batch)
+            response = self._make_request(
+                payload,
+                f"dataset/{dataset_id}/annotate"
+            )
+            if STATUS_CODE_KEY in response:
+                agg_response[ERRORS_KEY] = response
+            else:
+                agg_response[ANNOTATIONS_PROCESSED_KEY] += response[
+                        ANNOTATIONS_PROCESSED_KEY
```
Comment (rkaplan) [inline nucleus/__init__.py:388]: nit: whitespace should follow PEP8 here and next line

File: .vscode/settings.json [hunk 3]
```diff
@@ -0,0 +1,3 @@
+{
+    "python.pythonPath": "/Users/srikanthsrinivas/.pyenv/shims/python"
```
Comment (sasha-scale) [inline .vscode/settings.json:2]: This file should not be included in the PR

File: nucleus/annotation.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+            "type": "box",
+            "geometry": {
+                "x": self.x,
+                "y": self.y,
+                "width": self.width,
+                "height": self.height
+            },
+            "reference_id": self.reference_id,
+            "metadata": self.metadata
+        }
```
Comment (sasha-scale) [inline nucleus/annotation.py:24]: nit - add a newline at the end of this file to get rid of the stop sign thing

File: nucleus/dataset.py [hunk 5]
```diff
@@ -6,11 +10,30 @@ class Dataset:
...
 
     def __init__(self, dataset_id: str, client):
-        self.dataset_id = dataset_id
+        self.id = dataset_id
         self._client = client
 
     @property
-    def info(self) -> dict:
+    def name(self) -> str:
+        return self._info().get("name", "")
```
Comment (sasha-scale) [inline nucleus/dataset.py:18]: All of these string literals should use constants instead, I believe many of them already exist in the constants.py file

File: nucleus/errors.py [hunk 6]
```diff
@@ -0,0 +1,20 @@
+class ModelCreationError(Exception):
+    def __init__(self, message="Could not create the model"):
+        self.message = message
+        super().__init__(self.message)
+
+
```
Comment (sasha-scale) [inline nucleus/errors.py:6]: nit - extra newline here

File: nucleus/payload_constructor.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+from .annotation import BoxAnnotation
+from .prediction import BoxPrediction
+
+def construct_append_payload(dataset_items: List[DatasetItem]) -> dict:
+    payload = {}
+    items = []
+    for item in dataset_items:
+        items.append(item.to_payload())
+
+    return {"items": items}
```
Comment (sasha-scale) [inline nucleus/payload_constructor.py:12]: use constants instead of string literals here as well :)

**Review by sasha-scale — CHANGES_REQUESTED** [review]
Stylistically it looks pretty good to me, I haven't had a chance to check out and test yet, but I will absolutely do so later today

File: nucleus/__init__.py [hunk 8]
```diff
@@ -155,8 +188,7 @@ def delete_dataset(self, dataset_id: str) -> dict:
     def populate_dataset(
         self,
         dataset_id: str,
-        payload: dict,
-        local: bool = False,
+        dataset_items: List[DatasetItem],
         batch_size: int = 20,
```
Comment (sasha-scale) [inline nucleus/__init__.py:192]: missing "force" parameter, which is important if someone wants to overwrite datasetItems

Comment (sasha-scale) [reply nucleus/__init__.py:192]: Need to update doc string

File: nucleus/__init__.py [hunk 9]
```diff
@@ -294,17 +354,41 @@ def exception_handler(request, exception):
...
 
+        print("Remote upload:", async_responses)
         return async_responses
 
-    def annotate_dataset(self, dataset_id: str, payload: dict):
-        # TODO batching logic if payload is too large
+    def annotate_dataset(self, dataset_id: str, annotations: List[DatasetItem], batch_size: int = 20):
         """
         Uploads ground truth annotations for a given dataset.
         :param payload: {"annotations" : List[Box2DAnnotation]}
```
Comment (sasha-scale) [inline nucleus/__init__.py:363]: update docstring

**Review by sasha-scale — COMMENTED** [review]
Tested locally, was pretty intuitive to figure out. 

An additional improvement is that for items upload the user must convert to json to see the result:

`response.json()`

Whereas for annotations/predictions you simply do `print(response)`.  It would be great to standardize this.

Will also have to think about how to communicate to existing users that they need to install this update

File: nucleus/dataset_item.py [hunk 10]
```diff
@@ -0,0 +1,53 @@
+import os.path
+from .errors import FileNotFoundError
+from .constants import (
+    IMAGE_URL_KEY,
+    METADATA_KEY,
+    REFERENCE_ID_KEY,
+)
+
+
+class DatasetItem:
```
Comment (sasha-scale) [inline nucleus/dataset_item.py:10]: @srikanth-scale FYI for docs, I added optional field for item_id here.  Best way to instantiate a DatasetItem now is:

`DatasetItem(s3://my-bucket/picture.jpg, reference_id = 'my_pic', metadata = {...})`

File: nucleus/model.py [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+class Model:
+
+    def __init__(self, model_id: str, name: str, reference_id: str, metadata: dict, client):
+        self.id = model_id
+        self.name = name
+        self.reference_id = reference_id
+        self.metadata = metadata
+        self._client = client
+
+    def create_run(self, name: str, metadata: dict, dataset: Dataset, predictions: List[BoxPrediction]) -> ModelRun:
```
Comment (sasha-scale) [inline nucleus/model.py:16]: prob can delete this method?

File: nucleus/model_run.py [hunk 12]
```diff
@@ -58,20 +59,18 @@ def commit(self, payload: Optional[dict] = None) -> dict:
             payload = {}
         return self._client.commit_model_run(self.model_run_id, payload)
 
-    def predict(self, payload: dict) -> dict:
+    def predict(self, annotations: List[Any]) -> dict:
```
Comment (sasha-scale) [inline nucleus/model_run.py:62]: @srikanth-scale FYI for docs, updated this method to take list of 2DBoxPredictions instead of raw JSON payload.

File: nucleus/dataset_item.py [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+    def _is_local_path(self, path: str) -> bool:
+        path_components = [comp.lower() for comp in path.split("/")]
+        return not (
+            "https:" in path_components
+            or "http:" in path_components
+            or "s3:" in path_components
+        )
+
+    def _local_file_exists(self, path: str):
+        return os.path.isfile(path)
```
Comment (Nastia39) [inline nucleus/dataset_item.py:47]: I think, there methods should be moved to `utils` file as there aren't a part of `DatasetItem` abstraction

File: nucleus/model.py [hunk 14]
```diff
@@ -0,0 +1,28 @@
+from .dataset import Dataset
+from .prediction import BoxPrediction
+from .model_run import ModelRun
+from .payload_constructor import construct_box_predictions_payload
+from typing import List
+
+class Model:
+
+    def __init__(self, model_id: str, name: str, reference_id: str, metadata: dict, client):
```
Comment (Nastia39) [inline nucleus/model.py:9]: `metadata: Optional[dict]`

File: nucleus/model_run.py (hunk 12 above)
Comment (Nastia39) [reply nucleus/model_run.py:62]: `annotations: List[Union[Box2DPrediction, PolygonPrediction]]`

File: nucleus/dataset.py [hunk 15]
```diff
@@ -50,9 +83,11 @@ def create_model_run(self, payload: dict):
           "model_run_id": str,
         }
         """
-        return self._client.create_model_run(self.dataset_id, payload)
+        return self._client.create_model_run(self.id, payload)
 
-    def annotate(self, payload: dict) -> dict:
+    def annotate(
+        self, annotations: List[BoxAnnotation], batch_size=20
```
Comment (Nastia39) [inline nucleus/dataset.py:89]: nit: `batch_size: int = 20`

File: nucleus/dataset.py [hunk 16]
```diff
@@ -21,7 +54,7 @@ def info(self) -> dict:
                 'slice_ids': List[str]
             }
         """
-        return self._client.dataset_info(self.dataset_id)
+        return self._client.dataset_info(self.id)
 
     def create_model_run(self, payload: dict):
```
Comment (Nastia39) [inline nucleus/dataset.py:59]: change this method to a new style as well

File: nucleus/annotation.py [hunk 17]
```diff
@@ -0,0 +1,35 @@
+class BoxAnnotation:
+    def __init__(
+        self,
+        reference_id: str,
+        label: str,
+        x: int,
+        y: int,
+        width: int,
+        height: int,
+        metadata: dict = None,
```
Comment (Nastia39) [inline nucleus/annotation.py:10]: `metadata: Optional[dict] = None`

Comment (Nastia39) [reply nucleus/annotation.py:10]: Also add class for `PolygonAnnotation`

File: nucleus/__init__.py [hunk 18]
```diff
@@ -337,7 +466,15 @@ def add_model(self, payload: dict) -> dict:
         }
         :return: { "model_id": str }
```
Comment (Nastia39) [inline nucleus/__init__.py:467]: upd docs

**Review by Nastia39 — CHANGES_REQUESTED** [review]

**Review by jihan-yin — COMMENTED** [review]
Left some nits

File: nucleus/prediction.py [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+
+    def __str__(self):
+        return str(self.to_payload())
+
+
+class PolygonPrediction(PolygonAnnotation):
+    def __init__(
+        self,
+        label: str,
+        vertices: List[Any],
```
[the remark this answers is not in the document]
Comment (sasha-scale) [reply nucleus/prediction.py:66]: not yet lol, polygon annotation was a bit of an afterthought on this PR, unfortunately.  Left it as a TODO for the future :) good suggestion!

File: nucleus/model.py [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+            "reference_id": self.reference_id,
+            "metadata": metadata,
+        }
+        model_run: ModelRun = self._client.create_model_run(
+            dataset.id, payload
+        )
+
+        model_run.predict(predictions)
+
+        return model_run
```
[the remark this answers is not in the document]
Comment (sasha-scale) [reply nucleus/model.py:46]: I'm worried that adding a model_run as a property will be a bit misleading, because we don't automatically pull in all of the model_runs associated with this model (other runs could have been created previously).  I think the problem you're pointing out here is super valid though: how does a user fetch all model_runs associated with a model?  In the future we should add API endpoints for:
- get_model
- delete model
- get_model_runs for model

File: nucleus/annotation.py [hunk 21]
```diff
@@ -0,0 +1,110 @@
...
+    # pylint: disable=too-many-instance-attributes
+    def __init__(
+        self,
+        label: str,
+        vertices: List[Any],
+        reference_id: str = None,
+        item_id: str = None,
+        metadata: Optional[Dict] = None,
+    ):
+        if bool(reference_id) == bool(item_id):
```
[the remark this answers is not in the document]
Comment (sasha-scale) [reply nucleus/annotation.py:86]: As things stand now we actually require that exactly one of these fields be defined at time of upload.  I agree this is pretty suboptimal, I think in the future we should remove this and implement precendence between the two ids instead :) 

i.e. reference_id trumps dataset_item_id

**Review by sasha-scale — APPROVED** [review]
Let's ship it.

File: nucleus/model_run.py [hunk 22]
```diff
@@ -58,45 +60,57 @@ def commit(self, payload: Optional[dict] = None) -> dict:
...
         """
+        payload: Dict[str, List[Any]] = construct_box_predictions_payload(
+            annotations
+        )
         return self._client.predict(self.model_run_id, payload)
 
-    def iloc(self, i: int) -> dict:
+    def iloc(
+        self, i: int
+    ):  # TODO -> List[Union[BoxPrediction, PolygonPrediction]]:
```
Comment (Nastia39) [inline nucleus/model_run.py:83]: I added this to fix typing issue, have to be removed

File: nucleus/model_run.py [hunk 23]
```diff
@@ -58,45 +60,57 @@ def commit(self, payload: Optional[dict] = None) -> dict:
...
         }
         """
-        return self._client.predictions_iloc(self.model_run_id, i)
+        response = self._client.predictions_iloc(self.model_run_id, i)
+        return self._format_prediction_response(response)
 
-    def refloc(self, reference_id: str) -> dict:
+    def refloc(
+        self, reference_id: str
+    ):  # TODO -> List[Union[BoxPrediction, PolygonPrediction]]:
```
Comment (Nastia39) [inline nucleus/model_run.py:97]: I added this to fix typing issue, have to be removed [2]

**Review by Nastia39 — CHANGES_REQUESTED** [review]
